Sites Management Page: add support for promoting tailored flows

mpkelly opened the pull request:
Related to https://github.com/Automattic/wp-calypso/issues/68018

#### Proposed Changes

Show a banner component to promote Link in Bio sites. 

- [x] visuals match the proposed designs https://github.com/Automattic/wp-calypso/issues/68018#issuecomment-1257483200
- [x] clicking on button opens up the link-in-bio tailored flow
- [x] new tracks even for button clicks
- [x] promotion isn't shown if the user already has a Link in Bio site
- [x] promotion is dismissable
- [x] support RTL 
- [x] automated tests
- [x] manual tests for different screen widths and browsers

There are some things to note:

1. I track banner views because we seem to do that mostly elsewhere. Let me know if that is unnecessary. 
2. The tracks events are minimal - do we want more props sent up?
3. I matched the static Figma design quite closely. However, here are some differences:
  - The size of the tiles changes according to the available width, so it won't always look exactly like the mockup. 
  - I used the `Button` from `@wordpress/components`, which [has a height](https://wordpress.github.io/gutenberg/?path=/docs/components-button--buttons) of `36px` while the mockup uses `40px`. 
  - The images were taken from Figma but might need their size changes. Some are `3x` while others are `2x` (or close). I still need to add RTL images also. 
  - I would like to use `var(...)` where possible but have used values directly from Figma for now. 

I added a temp URL parameter to control the number of sites shown for testing and undo the dismiss. Just append `sitecount=1` to the URL and refresh. 

Example: http://calypso.localhost:3000/sites?sitecount=1

### Screenshots
![image](https://user-images.githubusercontent.com/6851384/193256486-7804495e-8cce-48c7-91d9-fbb3825e6963.png)

![image](https://user-images.githubusercontent.com/6851384/193256661-6ee81eb1-b2ef-4b62-b800-dc2397bb1248.png)

![image](https://user-images.githubusercontent.com/6851384/193256844-8841c85f-70d4-4b84-b60a-782e589b7ee8.png)

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Try it with one site `<live url>/sites?sitecount=1`
* Try it with two sites `<live url>/sites?sitecount=2`
* Try it with three sites `<live url>/sites?sitecount=3`
* Try it with one or two sites with an `rtl` language like Hebrew. 
* Try it with mobile
* Try it with various browser window widths from small through to large


#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #

mpkelly committed:
Link in Bio Banner: move rendering logic up into hook; improve UI on smaller screens & always use Tile banner on mobile.

p-jackson commented on client/sites-dashboard/components/link-in-bio-banner/BannerType.ts:
```diff
@@ -0,0 +1,6 @@
+export enum BannerType {
```
We try to avoid using TypeScript enums because they behave a little strange p7H4VZ-3sK-p2.

In this case I'd use a union type whenever I needed to reference all the banner types together.
```typescript
export type BannerType = 'none' | 'tile' | 'double-tile' | 'row';
```

Personally I don't think it's necessary, but if you prefer having named constants you'd just make  a constant for each one i.e. `export const BANNER_TYPE_NONE = 'none'`

mpkelly replied:
I will change that then. Thanks.

p-jackson commented on client/sites-dashboard/components/sites-table.tsx:
```diff
@@ -54,6 +55,13 @@ const Row = styled.tr`
 export function SitesTable( { className, sites, isLoading = false }: SitesTableProps ) {
+	// TODO just for testing the PR - will be removed.
+	const params = new URLSearchParams( window.location.search );
+	if ( params.get( 'sitecount' ) ) {
```
Thanks for adding the easy way to test the banners! 🎉

p-jackson commented on client/sites-dashboard/components/sites-grid.tsx:
```diff
@@ -29,13 +30,21 @@ interface SitesGridProps {
+	if ( params.get( 'sitecount' ) ) {
+		sites = sites.slice( 0, Number( params.get( 'sitecount' ) ) );
+	}
+	const { getBanner } = useLinkInBioBanner();
+	const banner = getBanner( sites, 'grid' );
```
The `useLinkInBioBanner` hook is used to render elements. To me it feels like a component disguised as a hook. Maybe it should be something more like:

```typescript
<LinkInBioBanner sites={ sites } displayMode="grid" />
```

I see the hook also has some shared functions, but I think those can be handled in another way.

mpkelly replied:
Yes, good point. That will be cleaner.

p-jackson commented on client/sites-dashboard/components/link-in-bio-banner/use-link-in-bio-banner.tsx:
```diff
@@ -0,0 +1,57 @@
+		dispatch( savePreference( preference, false ) );
+	};
+
+	const handleBannerViewed = () => {
+		recordTracksEvent( 'calypso_link_in_bio_banner_viewed' );
+	};
+
+	const handleBannerCtaClick = () => {
+		recordTracksEvent( 'calypso_link_in_bio_banner_cta_click' );
+	};
```
These look like pure helper functions. They don't need to be in a hook because their implementation isn't changing based on props. I think it makes more sense for these to just be regular helper functions outside of a hook.

p-jackson replied:
It looks like this also fixes the `React Hook useEffect has a missing dependency` lint warnings in the banner components. I guess the linter is smart enough to realise that the dependcy list isn't necessary when it's a function with static scope.

mpkelly replied:
I moved these up.

mpkelly committed:
Remove temporary test code!

p-jackson commented on client/sites-dashboard/components/link-in-bio-banner/use-link-in-bio-banner.tsx:
```diff
@@ -0,0 +1,57 @@
+		return Banners[ bannerType ];
+	};
+
+	// TODO just for testing the PR - will be removed.
```
Another handy way to reset preferences during development is to delete it using the Dev menu.

![CleanShot 2022-10-03 at 17 32 44@2x](https://user-images.githubusercontent.com/1500769/193502216-9bcb0270-a4e4-40a9-9a5c-1c84c8822f78.png)

mpkelly replied:
Noted, thanks.

p-jackson commented on client/sites-dashboard/components/link-in-bio-banner/use-link-in-bio-banner.tsx:
```diff
@@ -0,0 +1,57 @@
+		}
+	}, [] );
+
+	const handleDismissBanner = () => {
+		recordTracksEvent( 'calypso_link_in_bio_banner_dismiss_click' );
+		dispatch( savePreference( preference, false ) );
+	};
```
This function is only ever used in `<DismissButton>` so I think it makes more sense to inline the code there rather than have the hook return it.

mpkelly replied:
This was inlined. Thanks.

p-jackson commented on client/sites-dashboard/components/link-in-bio-banner/link-in-bio-banner.tsx:
```diff
@@ -0,0 +1,62 @@
+	margin: 0,
+} );
+
+Description.defaultProps = {
+	children: __( 'Show the world what you have to offer with a Link to Bio site.' ),
+};
```
I haven't seen this way of setting the text in a styled paragraph before! Convenient!
This uses the `__` function imported directly via `@wordpress/i18n`, but when we're dealing with React we usually use the `__` function that's returned by `@wordpress/react-i18n`'s `useI18n()` hook. In theory the React version means the component is able to re-render if the language ever changed. Although we never actually do that in practice—changing language always request a page refresh in Calypso.

mpkelly replied:
Ok, will switch over to the hook.

mpkelly replied:
After posting, I realised it wouldn't be possible to use the hook in this object. I've opted to keep it as-is and depend on the full refresh. Let me know if that's going to be problematic, and I can switch to a component.

p-jackson replied:
Yeah it'd be a bit of a change having to add a wrapper to the component just so you can use the hook. I think it's fine to leave as is, I was just thinking out loud that I hadn't seen it done this way before with default `children` props 😅

p-jackson reviewed (commented):
Coming along nicely! I think I need to delete my link-in-bio site so that I can see the banner appear, but I've left some comments in the meantime.

p-jackson commented on client/sites-dashboard/components/link-in-bio-banner/link-in-bio-double-tile-banner.tsx:
```diff
@@ -0,0 +1,65 @@
+	},
+	'@media screen and ( max-width: 960px )': {
+		padding: 32,
+		paddingRight: 0
```
This is where the CI lint error is coming from I think. Missing comma.

mpkelly replied:
I fixed that already when I switched to `paddingInlineEnd`, but thanks anyway.

mpkelly committed:
Link in Bio Banner: Add RTL support

mpkelly commented:
Regarding `rtl` support, I have added some simple rules, but maybe we should look at getting `@emotion` to do more for us as blocks like this as probably unnecessary:
```	
'html[dir=rtl] & ': {
     transform: 'scaleX(-1)', /* flip image */
},
```
There is some discussion [here](https://github.com/emotion-js/emotion/issues/793) in the `@emotion` repo.

p-jackson commented on client/sites-dashboard/components/link-in-bio-banner/link-in-bio-tile-banner.tsx:
```diff
@@ -14,6 +14,10 @@ const Root = styled( Banner.Root )( {
 		right: 16,
+		'html[dir=rtl] & ': {
+			right: 'unset',
+			left: 16,
```
Same as you've done in this commit, we've been using the "logical properties" CSS properties to support RTL. I also wish that emotion did a bit more for us in terms of RTL support though. The auto RTL tooling we use for the rest of Calypso (non-Emotion code) kinda just works by magic.

TIL there are logical properties for absolute positioning too. You can use `inset-inline-end` instead of `right`. https://css-tricks.com/building-multi-directional-layouts/#aa-logical-positioning-properties

mpkelly replied:
>TIL there are logical properties for absolute positioning too. You can use inset-inline-end instead of right. https://css-tricks.com/building-multi-directional-layouts/#aa-logical-positioning-properties

Ah! I did a quick search but didn't find it. Thanks!

mpkelly committed:
Link in Bio Banner: Address review comments

mpkelly commented:
> I think I need to delete my link-in-bio site so that I can see the banner appear, but I've left some comments in the meantime.

Actually, there is no logic (at least not from me) that would stop the banner from being shown. I still need to implement that requirement

FYI I addressed most of the review comments by accepting your suggestions.

mpkelly committed:
Add some unit tests

mpkelly commented:
Added some unit tests. I'd like to test mobile too but couldn't mock `matchMedia` even though I tried a bunch of ways including how we do it elsewhere in a `JSDOM` environment. Will get it working tomorrow when back online.

mpkelly committed:
Add mobile tests

mpkelly committed:
Not a *Linked* in Bio site...

mpkelly commented:
This is good for another review now. Just one thing, the unit tests failed due to a package duplication check. I don't think I introduce that and need to see if it's being looked at. The tests all pass though. 

```
[06:22:31]Step 4/11: Prevent duplicated packages (Command Line) (5s)
[06:22:31][Step 4/11] Running step within Docker container registry.a8c.com/calypso/base:latest
[06:22:31][Step 4/11] Starting: /bin/sh -c "docker pull registry.a8c.com/calypso/base:latest && . /home/teamcity-1/buildAgent/temp/agentTmp/docker-wrapper-18133346021007610289.sh && docker run --rm -w /home/teamcity-1/buildAgent/work/c4a9d5b38c1dacde --label jetbrains.teamcity.buildId=8724960 -u 1003 -v "/home/teamcity-1/buildAgent/lib:/home/teamcity-1/buildAgent/lib:ro" -v "/home/teamcity-1/buildAgent/tools:/home/teamcity-1/buildAgent/tools:ro" -v "/home/teamcity-1/buildAgent/plugins:/home/teamcity-1/buildAgent/plugins:ro" -v "/home/teamcity-1/buildAgent/work/c4a9d5b38c1dacde:/home/teamcity-1/buildAgent/work/c4a9d5b38c1dacde" -v "/home/teamcity-1/buildAgent/temp/agentTmp:/home/teamcity-1/buildAgent/temp/agentTmp" -v "/home/teamcity-1/buildAgent/temp/buildTmp:/home/teamcity-1/buildAgent/temp/buildTmp" -v "/home/teamcity-1/buildAgent/system:/home/teamcity-1/buildAgent/system" --env-file /home/teamcity-1/buildAgent/temp/agentTmp/docker-wrapper-3573505325895185962.envList --entrypoint /bin/sh "registry.a8c.com/calypso/base:latest" /home/teamcity-1/buildAgent/temp/agentTmp/docker-shell-script-10810900025842491886.sh"
[06:22:31][Step 4/11] in directory: /home/teamcity-1/buildAgent/work/c4a9d5b38c1dacde
[06:22:31][Step 4/11] latest: Pulling from calypso/base
[06:22:31][Step 4/11] Digest: sha256:07e6f9a9cbe7c0d3d3262246e827710f04b153b3fd8cf79d6fd0c2d6c323773e
[06:22:31][Step 4/11] Status: Image is up to date for registry.a8c.com/calypso/base:latest
[06:22:31][Step 4/11] registry.a8c.com/calypso/base:latest
[06:22:33][Step 4/11] Repository contains duplicated packages: 
[06:22:33][Step 4/11] 
[06:22:33][Step 4/11] ➤ YN0000: ┌ Deduplication step
[06:22:33][Step 4/11] ➤ YN0000: │ @types/yargs@npm:^17.0.1 can be deduped from @types/yargs@npm:17.0.5 to @types/yargs@npm:17.0.13
[06:22:33][Step 4/11] ➤ YN0000: │ One package can be deduped using the highest strategy
[06:22:33][Step 4/11] ➤ YN0000: └ Completed
[06:22:33][Step 4/11] 
[06:22:33][Step 4/11] To fix them, you need to checkout the branch, run 'yarn dedupe',
[06:22:33][Step 4/11] verify that the new packages work and commit the changes in 'yarn.lock'.
[06:22:33][Step 4/11] Process exited with code 1
[06:22:33][Step 4/11] Process exited with code 1 (Step: Prevent duplicated packages (Command Line))
```

sejas commented on client/sites-dashboard/components/sites-table.tsx:
```diff
@@ -115,36 +130,39 @@ export function SitesTable( { className, sites, isLoading = false }: SitesTableP
+					) ) }
+				</tbody>
+			</Table>
+			<LinkInBioBanner sites={ sites } displayMode={ 'row' } />
```
We can directly pass a string through props.

```suggestion
			<LinkInBioBanner sites={ sites } displayMode="row" />
```

sejas commented on client/sites-dashboard/components/link-in-bio-banner/link-in-bio-banner-parts.tsx:
```diff
@@ -0,0 +1,77 @@
+	return (
+		<Button
+			isPressed
+			href={ '/setup/intro?flow=link-in-bio&ref=logged-out-homepage-lp' }
```
We can directly pass a string through props.
```suggestion
			href="/setup/intro?flow=link-in-bio&ref=logged-out-homepage-lp"
```

sejas commented on client/sites-dashboard/components/link-in-bio-banner/link-in-bio-banner-parts.tsx:
```diff
@@ -0,0 +1,77 @@
+			icon={ close }
+			onClick={ handleDismissBanner }
+			{ ...props }
+			className={ 'dismiss-button' }
```
We can directly pass a string through props.

```suggestion
			className="dismiss-button"
```

sejas commented on client/sites-dashboard/components/link-in-bio-banner/link-in-bio-banner-parts.tsx:
```diff
@@ -0,0 +1,77 @@
+		<Button
+			isPressed
+			href={ '/setup/intro?flow=link-in-bio&ref=logged-out-homepage-lp' }
+			className={ 'create-button' }
```
```suggestion
			className="create-button"
```

sejas commented on client/sites-dashboard/components/link-in-bio-banner/link-in-bio-banner.tsx:
```diff
@@ -0,0 +1,55 @@
+	const showBanner =
+		( doesNotAlreadyHaveALinkInBioSite && isBannerVisible == null ) || isBannerVisible;
+
+	const renderBanner = () => {
+		let bannerType: BannerType = 'none';
+		if ( showBanner ) {
+			if ( isMobile ) {
+				bannerType = 'tile';
+			} else if ( displayMode === 'row' ) {
+				if ( siteCount === 1 ) {
+					bannerType = 'row';
+				}
+			} else if ( displayMode === 'grid' ) {
+				if ( siteCount === 1 ) {
+					bannerType = 'double-tile';
+				} else if ( siteCount === 2 ) {
+					bannerType = 'tile';
+				}
+			}
+		}
+
+		return LinkInBioBanners[ bannerType ];
+	};
+
+	return renderBanner();
```
I suggest removing the `renderBanner` function and returning the element directly. I also simplified the ifs a bit.

```suggestion
	let bannerType: BannerType = 'none';
	if ( showBanner ) {
		if ( isMobile ) {
			bannerType = 'tile';
		} else if ( displayMode === 'row' && siteCount === 1 ) {
			bannerType = 'row';
		} else if ( displayMode === 'grid' ) {
			if ( siteCount === 1 ) {
				bannerType = 'double-tile';
			} else if ( siteCount === 2 ) {
				bannerType = 'tile';
			}
		}
	}

	return LinkInBioBanners[ bannerType ];
```

mpkelly replied:
Yeah, the function wasn't adding anything here. Suggestion accepted.

sejas commented on yarn.lock:
```diff
@@ -7720,6 +7750,15 @@ __metadata:
   languageName: node
   linkType: hard
 
+"@types/yargs@npm:^17.0.8":
+  version: 17.0.13
+  resolution: "@types/yargs@npm:17.0.13"
+  dependencies:
+    "@types/yargs-parser": "*"
+  checksum: 5005e1e7408b9fea96c356becf19256f1a9bca02120852d3c0089ba7123528b0b6891185e9c92bac25cb5c04090a7a714b201523a6bf4a8a226852205c631208
+  languageName: node
+  linkType: hard
+
```
This is the error I'm seeing in team city:
```
Repository contains duplicated packages:

➤ YN0000: ┌ Deduplication step
➤ YN0000: │ @types/yargs@npm:^17.0.1 can be deduped from @types/yargs@npm:17.0.5 to @types/yargs@npm:17.0.13
➤ YN0000: │ One package can be deduped using the highest strategy
➤ YN0000: └ Completed

To fix them, you need to checkout the branch, run 'yarn dedupe',
verify that the new packages work and commit the changes in 'yarn.lock'.
```

mpkelly replied:
Thanks for that!

sejas reviewed (commented):
Tested, and this banner is looking good 🥳 .

I've added a few suggestions.

sejas commented on client/sites-dashboard/components/link-in-bio-banner/link-in-bio-double-tile-banner.tsx:
```diff
@@ -0,0 +1,67 @@
+		marginInlineStart: 'auto',
+		height: 173,
+		'html[dir=rtl] & ': {
+			transform: 'scaleX(-1)',
```
Nice trick flipping the image for rtl 👍.

sejas commented on client/sites-dashboard/components/link-in-bio-banner/link-in-bio-banner-parts.tsx:
```diff
@@ -0,0 +1,77 @@
+};
+
+export const CreateButton = () => {
+	return (
```
We could move the tracking inside the button.

Import `useEffect` and apply:
```suggestion
	useEffect( () => {
		recordTracksEvent( 'calypso_link_in_bio_banner_viewed' );
	}, [] );
	
	return (
```

Finally, remove the function `handleBannerViewed` and all its calls:
- https://github.com/Automattic/wp-calypso/blob/76c7c9497e40fc9a402931ed3cd20812dea3df09/client/sites-dashboard/components/link-in-bio-banner/link-in-bio-banner-parts.tsx#L75-L77
- https://github.com/Automattic/wp-calypso/blob/df99f2886e7f491f0ade4a6ffefa014d3d6b73bf/client/sites-dashboard/components/link-in-bio-banner/link-in-bio-double-tile-banner.tsx#L53
- https://github.com/Automattic/wp-calypso/blob/df99f2886e7f491f0ade4a6ffefa014d3d6b73bf/client/sites-dashboard/components/link-in-bio-banner/link-in-bio-row-banner.tsx#L26
- https://github.com/Automattic/wp-calypso/blob/df99f2886e7f491f0ade4a6ffefa014d3d6b73bf/client/sites-dashboard/components/link-in-bio-banner/link-in-bio-tile-banner.tsx#L33

Another option would be to refactor the function into a new custom hook `useBannerViewed`.

mpkelly replied:
Ok, I think that's worthwhile, so will make the change. 

One other thing, if the user toggles between list and and grid mode, the viewed stat will get incremented twice. Are we ok with this?

p-jackson replied:
That's an interesting point. Most of our "impression" or "viewed" tracks events have the same issue, they naively record an event on mount. So I'd say it's fine because it's a common issue for all of these sorts of events.

I don't think it generally causes an issue because if you use the Funnels tool (PCYsg-5hd-p2) it will only take notice of the first event anyway.

sejas reviewed (commented):
We can simplify the `calypso_link_in_bio_banner_viewed` logic.

sejas commented on client/sites-dashboard/components/link-in-bio-banner/link-in-bio-banner-parts.tsx:
```diff
@@ -0,0 +1,77 @@
+export const CreateButton = () => {
+	return (
+		<Button
+			isPressed
```
To improve the A11y, we should add a hover effect.
I've seen that the prop `isPressed` adds the class `is-pressed` to the button, which adds the black background. But this removes the hover effect, and it seems semantically incorrect.
I think we should remove the `isPressed` prop and style the button, adding the `hover` effect too.

mpkelly replied:
Good feedback. It feels like we should have an off-the-shelf `Button` in this style. Where would be an appropriate place to add such a `Button`? For now, I will just style a `Button` to match the design. 

Edit. We have a Link in Bio`Button` already that we use at `/setup/intro?flow=link-in-bio`. I've matched the style of that.

sejas replied:
@mpkelly 

Yes, It would be great if we had a styled button out of the box.
I've checked the storybook and devdocs, but none of those are using our colors. So it's totally ok to restyle them.

- https://wordpress.github.io/gutenberg/?path=/story/components-button--primary
- https://wpcalypso.wordpress.com/devdocs/design/button

The text has the same color as the hover background and it's not readable:
![bio-banner-hover](https://user-images.githubusercontent.com/779993/194019796-897f18e6-c812-4f89-b26c-d9b681d4e3bf.gif)

sejas reviewed (dismissed):
We should add the hover effect to the `CreateButton`.

mpkelly committed:
Update client/sites-dashboard/components/sites-table.tsx

Co-authored-by: Antonio Sejas <antonio@sejas.es>

mpkelly committed:
Update client/sites-dashboard/components/link-in-bio-banner/link-in-bio-banner-parts.tsx

Co-authored-by: Antonio Sejas <antonio@sejas.es>

mpkelly committed:
Update client/sites-dashboard/components/link-in-bio-banner/link-in-bio-banner-parts.tsx

Co-authored-by: Antonio Sejas <antonio@sejas.es>

mpkelly committed:
Update client/sites-dashboard/components/link-in-bio-banner/link-in-bio-banner.tsx

Co-authored-by: Antonio Sejas <antonio@sejas.es>

mpkelly committed:
Update client/sites-dashboard/components/link-in-bio-banner/link-in-bio-banner-parts.tsx

Co-authored-by: Antonio Sejas <antonio@sejas.es>

mpkelly committed:
Update lock file after dedupe

mpkelly committed:
Restyle button to make it like Button from setup page. Move handleBannerViewed in CreateButton

p-jackson commented:
> This is good for another review now
@mpkelly feel free to move PRs to the "Needs review" column on the board. It can help help get more eyes on it (I usually just want that column)

p-jackson commented on client/sites-dashboard/components/link-in-bio-banner/link-in-bio-banner.tsx:
```diff
@@ -0,0 +1,49 @@
+	return Boolean(
+		sites.find( ( site ) => {
+			const option = site.options || {};
+			return option.site_intent === 'link-in-bio';
```
We have access to modern JS in Calypso so we would usually do this like 
```suggestion
			return site.option?.site_intent === 'link-in-bio';
```

Just a nitpick, not required 😄

mpkelly replied:
For simple null-safe code where the thing that can be null is a leaf object, I still prefer `|| {}`. Where `?.` shines is for complex objects with many potentially null intermediate objects. Using `?.` adds more code to your final bundle. Here's a comparison from the [Babel REPL](https://babeljs.io/repl/#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=MYewdgzgLgBNCGUCmMC8MDeAoGMQAcoBLcCGALm11xxgF8sGtRJYDjS04pEkA6diUgwAPiMx0A3M1KwAbvAA2AVxTpBpPlAAWRMAHNJQA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Creact%2Cstage-2&prettier=false&targets=&version=7.19.3&externalPlugins=&assumptions=%7B%7D):

![image](https://user-images.githubusercontent.com/6851384/194195625-a42b04ad-9caa-488e-b4f3-513040af18a1.png)

![image](https://user-images.githubusercontent.com/6851384/194195634-38419c40-3f38-4a6c-935b-f2310cf7a523.png)

p-jackson replied:
> Using `?.` adds more code to your final bundle

That's a fair point. I'm looking forward to the day when our evergreen bundle won't need to do this!

p-jackson commented on client/sites-dashboard/components/link-in-bio-banner/test/link-in-bio-banner.tsx:
```diff
@@ -0,0 +1,89 @@
+import { LINK_IN_BIO_BANNER_PREFERENCE } from '../link-in-bio-banner-parts';
+
+describe( '<LinkInBioBanner>', () => {
+	each( [
```
I'm not a fan of including a new dependency for this. Usually when I see code for generating tests like this Calypso we just do it in a loop.

It should be possible to do this using the array's built in `forEach` method.

```typescript
	[
		//visible and right amount of rows - show
		[ true, 1, 'row', 'row-banner', true, false ],
		[ true, 1, 'grid', 'double-tile-banner', true, false ],

		// ... all the other test cases
	].forEach( ( [ visible, siteCount, type, output, expectValue, isMobile, siteIntent = 'other' ] ) => {
		test( `Parameters: visible: ${visible}, site count: ${siteCount}, type: ${type}, expectValue: ${expectValue}, isMobile: ${isMobile}, siteIntent: ${siteIntent}`, () => {
			if ( isMobile ) {
				// etc. etc.
		} );
	} );
```

You'd be able to remove the eslint ignore lines then too.

zaguiini replied:
Don't we have `test.each` included by default from Jest?

p-jackson replied:
TIL
https://jestjs.io/docs/api#testeachtablename-fn-timeout

Looks like Jest stole `jest-each`'s lunch!

mpkelly replied:
Ok, thanks for that. Let me switch it to `.each`. I thought I tried that first, but maybe I didn't use the correct syntax.

p-jackson commented on client/sites-dashboard/components/link-in-bio-banner/link-in-bio-double-tile-banner.tsx:
```diff
@@ -0,0 +1,64 @@
+	padding: '0 64px',
+	paddingInlineEnd: 0,
+	height: '100%',
+	'.dismiss-button': {
```
For context: we've only recently started using `@emotion`—the sites management page is our first project using it. So we're still figuring out the patterns we like.

Fwiw we haven't been using nested class selectors like this in Emotion. It's perfectly valid CSS but I guess it works around the purpose of Emotion: defining your own class names and breaking that strong component-level style encapsulation. Given that dismiss button styles for both the single and double tile banner are identical it probably makes sense to define those styles as part of the `DismissButton` component.

mpkelly replied:
>Fwiw we haven't been using nested class selectors like this in Emotion. It's perfectly valid CSS but I guess it works around the purpose of Emotion: defining your own class names and breaking that strong component-level style encapsulation

I totally agree! I haven't used `@emotion` much and was on the `styled-components` side instead. Here's what you can do in `styled-components` to reference children:

```js
[ `${ Banner.DismissButton }` ]: { // dynamic class name gets inserted here
    position: 'absolute',
    top: 16,
    insetInlineEnd: 16,
},
```
This doesn't work with `@emotion`, and the best I found was [here](https://www.matthewsessions.com/blog/emotion-nested-selectors/), which is a bit more cumbersome. Maybe there is a better way?

>Given that dismiss button styles for both the single and double tile banner are identical it probably makes sense to define those styles as part of the DismissButton component.

I tend to prefer having everything in one place rather than overriding stuff. Saving a few lines of code seems good but makes it harder for the next person. Now I only use base styles if they are common to all instances, but the `row` banner uses different styles and would need to override the base styles, which I think makes the code harder to read. 

Dan Abramov wrote an article about "clean code" [here](https://overreacted.io/goodbye-clean-code/). It's a good read. 

On the subject of `styled` libraries. It's difficult to overlook stitches in 2022. It has very little runtime overhead and great theming and variant support. 

https://stitches.dev/

p-jackson replied:
Cool! I hadn't heard of stitches. The marketing claims at least sound great 😄 
We went with Emotion because it's what's being used by core WordPress, so probably unlikely we'd go in a different direction from them.

Personally I haven't been fully convinced by css-in-js. My impression so far is that it's probably great for theming, and there are probably cool SSR things you can do with it e.g. only send the CSS rules you need, but it doesn't add enough over and above plain CSS in the contexts that I've used it.

p-jackson reviewed (commented):
The PNGs take a while to load for me. (I'm on 4G right now, which is good for testing!) 470KB seems quite big for this sized image! I wonder if JPEG would be more appropriate given they're photos. The downside is that JPEG doesn't support transparency.
You could bake the purple colour into the background maybe? Or perhaps use a CSS trick to add round corners to a square jpeg?

These are all optimisations for another PR though I think.

When the tile view drops down to two tiles per row in narrow windows (or on vertical iPad) it looks like this:
![ipad](https://user-images.githubusercontent.com/1500769/194016622-479eccf7-bac7-467e-bcfd-b6dc20e0aa29.png)

I don't think this was quite the intention of the design. IMO the banner looks best when it's tidying up an incomplete row, not when it's creating a whole new incomplete row. I don't know if there's a quick CSS fix. You could perhaps hide the banner when there are two sites maybe. That was actually something that was discussed in the task thread.

I think the issue mentioned above with hover state is the last remaining major issue. Everything else is working great!

For reference, this is what the button hover currently looks like
![hover](https://user-images.githubusercontent.com/1500769/194017554-521ba32f-94d9-49cb-a668-9766903e58eb.png)

zaguiini commented:
I think the button text should not wrap here (i.e. take precedence over the text and break the text on a new line):

![image](https://user-images.githubusercontent.com/26530524/194142731-2dff57a9-a8c4-4fd5-b6d5-0181bfb68865.png)

When there's one site, but not enough space for three tiles, I think the link-in-bio banner should be reduced to a single tile, otherwise, we'd see a blank tile to the right: 

![image](https://user-images.githubusercontent.com/26530524/194143758-dddb07e3-b445-44b1-8a4f-590f515e8fcb.png)

What do you think of adding a bottom link-in-bio banner if there are exactly 3 sites? It looks like we could make the most of that empty space there:

![image](https://user-images.githubusercontent.com/26530524/194143964-522df701-7cec-472a-b7fc-2fdc473d3592.png)

p-jackson commented:
> What do you think of adding a bottom link-in-bio banner if there are exactly 3 sites?

I'd prefer to focus on only showing the banner to users who have very few sites. The more sites they have I think the less nudging we need to do for them—they seem to know what they're doing. Users with multiple sites probably need different banner text.

mpkelly commented:
@p-jackson, regarding the hover styles, here's what I see on Chrome. How does this compare to what you're seeing? I am running this from my local Calypso environment. 

![image](https://user-images.githubusercontent.com/6851384/194197116-0688cee6-ebcd-4cd6-b7a7-3d3caa0e1a7d.png)

mpkelly committed:
Remove `jest-each` and use the built-in version.

mpkelly committed:
Don't shrink button CTA button when space is tight

mpkelly commented:
@zaguiini I made the easier change to the button text not being fully visible. 

![image](https://user-images.githubusercontent.com/6851384/194199932-e27b4f2e-a2bb-4374-a47f-f840289b83a3.png)

I agree with your other change but think it should be handled in a follow-up PR with some of the stuff @p-jackson mentioned.

p-jackson commented:
> How does this compare to what you're seeing? I am running this from my local Calypso environment.

Oh I bet the issue is to do with Calypso themes! I'm using the **Classic Blue** theme, and `var( --color-surface )` must look invisible in this theme.

mpkelly commented:
>Oh I bet the issue is to do with Calypso themes! I'm using the Classic Blue theme, and var( --color-surface ) must look invisible in this theme.

Thanks @p-jackson. I tried Classic Blue, and it's still ok for me. I tried a bunch of themes, and `color-surface` seems always to be `#fff`. I originally copied the styles from the `button` here at `/setup/intro?flow=link-in-bio&ref=logged-out-homepage-lp`. Do you also see the problem there?

Anyway, I checked Figma and the color `--studio-gray-0` is used. This sounds like a more stable variable name than `surface` so I switched to that.

mpkelly committed:
Switch button color to --studio-gray-0

mpkelly commented:
I think this is good for one final review. Based on previous reviews, a follow-up PR will be needed for

- Optimise images, so they load faster on slow networks
- Change banner type according to available space: `double-tile` should become `tile`. 
- Don't show the banner when it wraps onto a line by itself

p-jackson commented:
@mpkelly the hover issue is still happening for me 🙃 

![CleanShot 2022-10-07 at 15 04 53@2x](https://user-images.githubusercontent.com/1500769/194451138-ca511496-485f-4b41-a039-029a1c9243e7.png)

This `--color-neutral-70` is coming into play here I believe. https://github.com/Automattic/wp-calypso/blob/9ffd3524bd656fc3ee7f9756d1448cb03f5afcb3/packages/components/src/button/style.scss#L58
I don't know why I'd be seeing that and not you though. Maybe CSS include order? I'm testing on `container-magical-bardeen.calypso.live` if that's what's making the difference to CSS ordering.

> This branch has conflicts that must be resolved

I usually try to resolve these as I go, that way any reviewing and testing that's being done is also testing the final rebased code.

> follow-up PR will be needed for

I agree with following up on those 3 things

mpkelly committed:
Rebase and merge with trunk

mpkelly committed:
Rebase and merge with trunk

mpkelly committed:
Add color style to hover rule which is needed in some environments (e.g. Calypso Live)

mpkelly commented:
> @mpkelly the hover issue is still happening for me 🙃

@p-jackson  Yes, I checked Live and it was happening for me there too. It is fixed now along with the merge conflicts.

p-jackson commented on client/sites-dashboard/components/sites-table.tsx:
```diff
@@ -115,36 +130,39 @@ export function SitesTable( { className, sites, isLoading = false }: SitesTableP
+				>
+					<Row>
+						<th style={ { width: '50%' } }>{ __( 'Site' ) }</th>
+						<th style={ { width: '20%' } }>{ __( 'Plan' ) }</th>
+						<th>{ __( 'Last Publish' ) }</th>
+						<th>{ __( 'Status' ) }</th>
+						<th style={ { width: '24px' } }></th>
```
A typo looks like it's crept in somehow where the column headers have flipped the wrong way

<img width="294" alt="CleanShot 2022-10-07 at 16 20 03@2x" src="https://user-images.githubusercontent.com/1500769/194461043-480fb392-98e0-45aa-b20e-fd048ffc2d01.png">


```suggestion
						<th style={ { width: '20%' } }>{ __( 'Plan' ) }</th>
						<th>{ __( 'Status' ) }</th>
						<th>{ __( 'Last Publish' ) }</th>
						<th style={ { width: '24px' } }></th>
```

p-jackson commented on client/sites-dashboard/components/link-in-bio-banner/link-in-bio-banner-parts.tsx:
```diff
@@ -0,0 +1,97 @@
+		color: 'var(--studio-gray-0)',
+	},
+	'&:focus': {
+		backgroundColor: 'var(--studio-gray-70)',
+		outline: 'solid 2px var(--studio-gray-70)',
```
Setting the colour explicitly on hover has fixed the issue. Unfortunately the focus state has the same issue.

<img width="410" alt="CleanShot 2022-10-07 at 16 17 01@2x" src="https://user-images.githubusercontent.com/1500769/194461355-997357f1-2fb8-489f-8579-cce71c534fff.png">

But you should be able to just apply the same fix to it as well.

```suggestion
		backgroundColor: 'var(--studio-gray-70)',
		color: 'var(--studio-gray-0)',
		outline: 'solid 2px var(--studio-gray-70)',
```

p-jackson reviewed (approved):
This is looking good! I found a couple of trivial issues, but other than that this change is good to go 🚢 

Approved, but before merging:
- fix focus state
- fixup table heading order
- remove the query param code

mpkelly committed:
Update client/sites-dashboard/components/link-in-bio-banner/link-in-bio-banner-parts.tsx

Co-authored-by: Philip Jackson <p-jackson@live.com>

mpkelly committed:
Update client/sites-dashboard/components/sites-table.tsx

Co-authored-by: Philip Jackson <p-jackson@live.com>

mpkelly committed:
Remove test code for controlling site count

mpkelly commented:
Thanks for catching those issues and providing fixes, @p-jackson! I have removed the temp code and it's building now.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7568544

Thank you @mpkelly for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7568544).
